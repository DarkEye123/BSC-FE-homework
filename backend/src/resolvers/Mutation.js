import { forwardTo } from 'prisma-binding';
import signIn from './mutations/signIn';
import logout from './mutations/logout';
import createNote from './mutations/createNote';

const Mutation = {
  signIn,
  logout,
  createNote,
  deleteNote: forwardTo('prisma'),
  updateNote: forwardTo('prisma'),
};

export default Mutation;
