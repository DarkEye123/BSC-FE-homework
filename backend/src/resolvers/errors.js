/* eslint-disable no-proto */
class AppErr extends Error {
  constructor(msg = 'oopsie ... unknown error (╯°□°)╯') {
    super(msg);
    // this is useful for Jest and error subclassing https://github.com/facebook/jest/issues/5416#issuecomment-456414004
    this.constructor = AppErr;
    this.__proto__ = AppErr.prototype;
  }
}

class NotImplemented extends AppErr {
  constructor(msg = 'oopsie ... not implemented (╯°□°)╯') {
    super(msg);
    this.constructor = NotImplemented;
    this.__proto__ = NotImplemented.prototype;
  }
}

export { AppErr, NotImplemented };
