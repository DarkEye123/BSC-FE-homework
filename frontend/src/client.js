import ApolloClient from 'apollo-boost';
import fetch from 'node-fetch';

function createClient() {
  return new ApolloClient({
    uri: process.env.REACT_APP_BACKEND_URL,
    fetch,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
      });
    },
  });
}

export default createClient;
