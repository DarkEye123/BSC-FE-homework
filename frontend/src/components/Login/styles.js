import styled from 'styled-components';
import { Form as FormikForm } from 'formik';
import { InputField } from '../common/InputField/styles';

const Form = styled(FormikForm)`
  padding: 4rem;
  border: 1px solid ${({ theme }) => theme.color.lightShades}
  ${InputField}:first-child {
    margin-bottom: 2rem;
  }
  button {
    margin: 2rem 0 0 0;
    width: 100%;
  }
`;

export { Form };
