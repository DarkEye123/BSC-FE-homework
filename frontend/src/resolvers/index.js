export { default as LOGIN_MUTATION } from './mutations/login';
export { default as LOGOUT_MUTATION } from './mutations/logout';
export { default as DELETE_NOTE_MUTATION } from './mutations/deleteNote';
export { default as CREATE_NOTE_MUTATION } from './mutations/createNote';
export { default as ACTIVE_USER_QUERY } from './queries/activeuser';
export { default as NOTES_QUERY } from './queries/notes';
