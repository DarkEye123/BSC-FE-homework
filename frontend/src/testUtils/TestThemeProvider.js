import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import theme from '../theme';

const TestThemeProvider = ({ children }) => <ThemeProvider theme={theme}>{children}</ThemeProvider>;
TestThemeProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default TestThemeProvider;
