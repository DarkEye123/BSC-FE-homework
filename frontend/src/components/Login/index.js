import React from 'react';
import { Formik } from 'formik';
import { CenteredMain as Page, InputField, ErrorMessage, SpinningButton } from '..';
import { Mutation } from 'react-apollo';
import { withRouter } from 'react-router';
import { useTranslation } from 'react-i18next';
import { Form } from './styles';
import LoginSchema from './validations';
import { ACTIVE_USER_QUERY, LOGIN_MUTATION } from '../../resolvers';
import { PAGES } from '../../routes';

const update = (cache, { data: { signIn } }) => {
  const data = { activeUser: signIn };
  cache.writeQuery({ data, query: ACTIVE_USER_QUERY });
};

// Note: 'catch' statement is here due to red warnings made by graphql error response
const handleOnSubmit = (mutation, history) => async values => {
  try {
    await mutation({ variables: { email: values.email, password: values.password } });
    history.push(PAGES.notes);
  } catch (e) {}
};

// eslint-disable-next-line react/prop-types
const Login = ({ history }) => {
  const { t } = useTranslation();
  return (
    <Mutation mutation={LOGIN_MUTATION} update={update}>
      {(logIn, { loading, error }) => (
        <Formik
          initialValues={{ email: '', password: '' }}
          validationSchema={LoginSchema}
          onSubmit={handleOnSubmit(logIn, history)}
        >
          {({ isValid }) => (
            <Page data-testid="loginpage">
              <ErrorMessage error={error} />
              <Form>
                <InputField data-testid="email" label="Email" name="email" />
                <InputField data-testid="password" label="Password" name="password" />
                <SpinningButton data-testid="signin" disabled={!isValid || loading} spin={loading} type="submit">
                  {t(`Sign${loading ? 'ing' : ''} In`)}
                </SpinningButton>
              </Form>
            </Page>
          )}
        </Formik>
      )}
    </Mutation>
  );
};

export default withRouter(Login);
