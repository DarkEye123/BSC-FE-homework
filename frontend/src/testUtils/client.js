/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { MockLink } from 'apollo-link-mock';
import { ApolloProvider } from 'react-apollo';
import { ApolloProvider as ApolloHooksProvider } from 'react-apollo-hooks';
import TestThemeProvider from './TestThemeProvider';

function createClient(mocks) {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new MockLink(mocks),
  });
}

function createPage(client, children) {
  return (
    <ApolloProvider client={client}>
      <ApolloHooksProvider client={client}>
        <TestThemeProvider>{children}</TestThemeProvider>
      </ApolloHooksProvider>
    </ApolloProvider>
  );
}

export { createClient, createPage };
