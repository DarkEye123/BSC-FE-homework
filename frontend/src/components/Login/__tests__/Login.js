import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { cleanup, fireEvent, render, wait } from 'react-testing-library';
import Login from '..';
import { LOGIN_MUTATION, ACTIVE_USER_QUERY } from '../../../resolvers';
import { createClient, createPage } from '../../../testUtils/client';

const VALID_PASSWORD = '123qwer684a34sf';
const INVALID_PASSWORD = 'abrakadabra';
const EMAIL = 'test@tester.com';
const UID = '1234';

const mocks = [
  {
    request: {
      query: LOGIN_MUTATION,
      variables: { email: EMAIL, password: VALID_PASSWORD },
    },
    result: { data: { signIn: { id: UID, email: EMAIL, __typename: 'User' } } },
  },
  {
    request: {
      query: LOGIN_MUTATION,
      variables: { email: EMAIL, password: INVALID_PASSWORD },
    },
    error: new Error('shit happens'),
  },
];

const page = (history = {}, client = createClient(mocks)) =>
  // eslint-disable-next-line implicit-arrow-linebreak
  createPage(client, <MemoryRouter>{Login(history)}</MemoryRouter>);

afterEach(cleanup);

describe('login tests', () => {
  it('renders', () => {
    const { getByTestId } = render(page());
    expect(getByTestId('loginpage')).toMatchSnapshot();
  });

  it('validates', async () => {
    const history = { push: jest.fn() };
    const client = createClient(mocks);

    const { getByTestId } = render(page(history, client));
    const signInBtn = getByTestId('signin');
    const emailInput = getByTestId('email');
    const passwordInput = getByTestId('password');

    fireEvent.change(emailInput, { target: { value: 'omacka' } });
    fireEvent.change(passwordInput, { target: { value: VALID_PASSWORD } });
    await wait();
    expect(signInBtn).toBeDisabled();
  });

  it('can log in', async () => {
    const history = { push: jest.fn() };
    const client = createClient(mocks);

    const { getByTestId } = render(page(history, client));
    const signInBtn = getByTestId('signin');
    const emailInput = getByTestId('email');
    const passwordInput = getByTestId('password');

    fireEvent.change(emailInput, { target: { value: EMAIL } });
    fireEvent.change(passwordInput, { target: { value: VALID_PASSWORD } });
    await wait();
    fireEvent.click(signInBtn);

    await wait(() => {
      expect(client.cache.readQuery({ query: ACTIVE_USER_QUERY })).toEqual({
        activeUser: { id: UID, email: EMAIL, __typename: 'User' },
      });
    });
  });

  it('it shows BE error', async () => {
    const client = createClient(mocks);
    const { getByTestId } = render(page({}, client));
    const signInBtn = getByTestId('signin');
    const emailInput = getByTestId('email');
    const passwordInput = getByTestId('password');

    fireEvent.change(emailInput, { target: { value: EMAIL } });
    fireEvent.change(passwordInput, { target: { value: INVALID_PASSWORD } });
    await wait();
    fireEvent.click(signInBtn);

    await wait(() => {
      expect(getByTestId('graphql-error').textContent).toMatchInlineSnapshot('"Problem!Network error: shit happens"');
    });
  });
});
