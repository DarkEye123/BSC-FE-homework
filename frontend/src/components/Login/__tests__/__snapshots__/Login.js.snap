// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login tests renders 1`] = `
<main
  class="sc-bwzfXH frOdhr"
  data-testid="loginpage"
>
  <form
    class="sc-kGXeez dYYneR"
  >
    <div
      class="sc-bZQynM cCBhlA"
    >
      <input
        class="sc-gzVnrw kuqIpM"
        data-testid="email"
        id="email"
        name="email"
        value=""
      />
      <span />
      <label
        data-testid="email"
        for="Email"
      >
        Email
      </label>
    </div>
    <div
      class="sc-bZQynM cCBhlA"
    >
      <input
        class="sc-gzVnrw kuqIpM"
        data-testid="password"
        id="password"
        name="password"
        value=""
      />
      <span />
      <label
        data-testid="password"
        for="Password"
      >
        Password
      </label>
    </div>
    <button
      class="sc-dnqmqq bsujTy"
      data-testid="signin"
      disabled=""
      type="submit"
    >
      Sign In
    </button>
  </form>
</main>
`;
